[quant] Add quantize and dequantize operators to decomposition table

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93312

Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant] decompose quantize and dequantize operators

Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qconfig_module_type

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

leslie-fang-intel commented:
Not sure I am understanding it right? after this PR merged, we will not see fx node of `torch.ops.quantized_decomposed.quantize_per_tensor` but more primitive aten operators after `convert_pt2e`. It may make `dq-conv-q` pattern match and fusion more complicated such as in `inductor` after `convert_pt2`. cc @jgong5 @Xia-Weiwen

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qconfig_module_type

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> Not sure I am understanding it right? after this PR landed, we will not see fx node of `torch.ops.quantized_decomposed.quantize_per_tensor` but more primitive aten operators after `convert_pt2e`. It may make `dq-conv-q` pattern match and fusion more complicated such as in `inductor` after `convert_pt2`. cc @jgong5 @Xia-Weiwen

this is not enabled by default right now, but the decomposed representation might be the default, and we are still discussing whether we want to expose the representation with q/dq node or not

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qconfig_module_type

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/_decomp/decompositions.py:
```diff
@@ -2761,6 +2762,33 @@ def upsample_bicubic2d_vec(
 
 
+@register_decomposition(torch.ops.quantized_decomposed.quantize_per_tensor)
+def quantize_per_tensor(
```
n00b question, what's the difference between the original function and this function?  They both seem to do the same thing, with the math in Python.  Are there some systems that treat functions in `torch/_decomp/decompositions.py` differently?  Note: I've never used this file.

jerryzh168 replied:
I think they serve different purposes, the original function is for execution, the function in decomposition is for generating the decomposition for a specific op, although maybe we can just use a single implementation

SherlockNoMad commented on torch/_decomp/decompositions.py:
```diff
@@ -19,6 +19,7 @@
     _safe_copy_out,
     out_wrapper,
 )
+from torch.ao.quantization.fx._decomposed import quantized_decomposed_lib  # noqa: F401
```
I would suggest these decompositions for custom ops to live together with the operator definition. Something like ao/quantization/decompositions.py ?

User should be able to define custom op, define decompositions for custom op, all in user script.

SherlockNoMad replied:
Let me know if this doesn't work... 
My goal is to enable dynamic decomposition registration for custom op.

SherlockNoMad reviewed (changes requested):
Per comment.

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_pt2e/decompositions.py:
```diff
@@ -0,0 +1,29 @@
+from torch._decomp import register_decomposition
+
+@register_decomposition(torch.ops.quantized_decomposed.quantize_per_tensor)
+def quantize_per_tensor(
```
@SherlockNoMad one question I have is that if I should just reuse the functions defined in `torch.ao.quantization.fx._decmposed`, these functions have the same impl, but they have some additional checks

SherlockNoMad replied:
Yes, looks very reasonable to me to just reuse.

jerryzh168 replied:
but then the additional checks will be traced as well? can we remove them?

vkuzo replied:
sorry, I'm just ramping on on decomps.  I think we should avoid copy-pasta here.  The naive solution would be to leave the checks in `torch/ao/quantization/fx/_decomposed.py` and have that function call here for the math.

jerryzh168 replied:
sure, I can do that

SherlockNoMad commented on torch/_meta_registrations.py:
```diff
@@ -2698,6 +2698,9 @@ def activate_meta():
             elif "mkl::" in op_overload.name():
                 _meta_lib_dont_use_me_use_register_meta_for_mkl.impl(op_overload, fn)
+            elif "quantized_decomposed::" in op_overload.name():
+                lib = torch.library.Library("quantized_decomposed", "IMPL", "Meta")
```
nit, maybe rewrite this to only load lib once... 
I don't know how expensive it is to create a lib every iteration.

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorch merge -f "failures unrelated to PR"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but it breaks trunk due to a landrace https://hud.pytorch.org/pytorch/pytorch/commit/3fd46a2f9c56c692b242727cb146cfd464210c6a.  Please rebase and re-land it' -c landrace

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but it breaks trunk due to a landrace https://hud.pytorch.org/pytorch/pytorch/commit/3fd46a2f9c56c692b242727cb146cfd464210c6a. Please rebase and re-land it' -c landrace

jerryzh168 commented:
@pytorchbot rebase

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jerryzh168 commented:
Hi @malfet and @albanD, please advise on how to land this PR, looks like the dynamo related tests are still failing: https://hud.pytorch.org/pr/93312

jerryzh168 committed:
Update on "[quant] Add quantize and dequantize operators to decomposition table"


Summary:
This PR tries to decompose the operators in torch.ops.quantized_decomposed namespace to more
primitive aten operators, this would free us from maintaining the semantics of the quantize/dequantize
operators, which can be expressed more precises in terms of underlying aten operators

Note: this PR just adds them to the decomposition table, we haven't enable this by default yet

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_q_dq_decomposition

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jeanschmidt commented:
@pytorchbot revert -m "this commits breaks internal builds: https://fburl.com/sandcastle/dw0rqcbv" -c nosignal

Xia-Weiwen commented:
Hi @jerryzh168 I think we found some issues with this PR. Running the following code with this PR, one will get an error.
Code:
```python
import torch
import torch._dynamo as torchdynamo
from torch.testing._internal.common_quantized import (
    override_quantized_engine,
)
from torch.ao.quantization import (
    get_default_qconfig,
    QConfigMapping,
)
from torch.ao.quantization.backend_config._qnnpack_pt2e import get_qnnpack_pt2e_backend_config
from torch.ao.quantization._quantize_pt2e import prepare_pt2e, convert_pt2e
import copy
from torch._inductor.compile_fx import compile_fx


class Mod(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        # self.relu = torch.nn.ReLU()

    def forward(self, x):
        # return self.relu(x + x)
        return x + x

with override_quantized_engine("qnnpack"):
    example_inputs = (torch.randn(1, 3, 224, 224),)
    m = Mod().eval()
    # program capture
    m, guards = torchdynamo.export(
        m,
        *copy.deepcopy(example_inputs),
        aten_graph=True,
        tracing_mode="real",
    )

    backend_config = get_qnnpack_pt2e_backend_config()
    qconfig = get_default_qconfig("qnnpack")
    qconfig_mapping = QConfigMapping().set_global(qconfig)
    before_fusion_result = m(*example_inputs)

    m = prepare_pt2e(m, qconfig_mapping, example_inputs, backend_config)
    m(*example_inputs)
    m = convert_pt2e(m)

    run = torch._dynamo.optimize(compile_fx, nopython=False)(m)
    inductor_result = run(*example_inputs)
```
Error:
```
Traceback (most recent call last):
  File "/home/weiwen/.conda/envs/int8-dev/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1193, in run_node
    return node.target(*args, **kwargs)
  File "/home/weiwen/.conda/envs/int8-dev/lib/python3.9/site-packages/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: quantized_decomposed::quantize_per_tensor() Expected a value of type 'float' for argument 'scale' but instead found type 'FakeTensor'.
Position: 1
Value: FakeTensor(FakeTensor(..., device='meta', size=()), cpu)
Declaration: quantized_decomposed::quantize_per_tensor(Tensor input, float scale, int zero_point, int quant_min, int quant_max, ScalarType dtype) -> Tensor
Cast error details: Unable to cast Python instance to C++ type (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```
As you can see the dispatching of `quantize_per_tensor` to 'meta' failed.

We think there are at least problems. First is that decomposed quantize and dequantize are not successfully registered to meta. This is because `activate_meta()` in `torch/_meta_registrations.py` is called when importing torch and at that time `_decomposed.py` is not imported yet. If you add a `print` here https://github.com/pytorch/pytorch/blob/a7b5896197d7acdcf19a88d233b99045dc0a7293/torch/_meta_registrations.py#L2705, you will find no op been registered.
The second problem is that the dispatcher somehow fonud the registration of `quantize_per_tensor` but not `quantize_per_tensor.tensor`. So, the types of arguments mismatched.
Could you take a look? Thanks!

Also cc @leslie-fang-intel
